tracer/src/Datadog.Trace/Debugger/Symbols/SymbolsUploader.cs

```diff
@@ -0,0 +1,289 @@
+            }
+
+            Log.Information("Symbol database uploading is disabled. To enable it, please set {EnvironmentVariable} environment variable to 'true'.", ConfigurationKeys.Debugger.SymbolDatabaseUploadEnabled);
+            return new NoOpUploader();
+        }
+
+        private void RegisterToAssemblyLoadEvent()
+        {
+            AppDomain.CurrentDomain.AssemblyLoad += async (_, args) =>
+            {
+                await ProcessItemAsync(args.LoadedAssembly).ConfigureAwait(false);
+            };
+        }
+
+        private async Task ProcessItemAsync(Assembly assembly)
+        {
+            await Task.Yield();
+            await _assemblySemaphore.WaitAsync(_cancellationToken.Token).ConfigureAwait(false);
+
+            if (_cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+
+            try
+            {
+                await ProcessItem(assembly).ConfigureAwait(false);
+            }
+            finally
+            {
+                _assemblySemaphore.Release();
+            }
+        }
+
+        private async Task ProcessItem(Assembly assembly)
+        {
+            try
+            {
+                if (AssemblyFilter.ShouldSkipAssembly(assembly))
+                {
+                    return;
+                }
+
+                if (!_alreadyProcessed.Add(assembly.Location))
+                {
+                    return;
+                }
+
+                await UploadAssemblySymbols(assembly).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Error while trying to extract assembly symbol {Assembly}", assembly);
+            }
+        }
+
+        private async Task UploadAssemblySymbols(Assembly assembly)
+        {
+            using var symbolsExtractor = SymbolExtractor.Create(assembly);
+            if (symbolsExtractor == null)
+            {
+                return;
+            }
+
+            var assemblyScope = symbolsExtractor.GetAssemblySymbol();
+
+            var root = new Root
+            {
+                Service = _serviceName,
+                Env = _environment,
+                Language = "dotnet",
+                Version = _serviceVersion,
+                Scopes = new[] { assemblyScope }
+            };
+
+            await UploadClasses(root, symbolsExtractor.GetClassSymbols()).ConfigureAwait(false);
+        }
+
+        private async Task UploadClasses(Root root, IEnumerable<Model.Scope?> classes)
+        {
+            var accumulatedBytes = 0;
+            var builder = StringBuilderCache.Acquire((int)_thresholdInBytes);
+
+            try
+            {
+                foreach (var classSymbol in classes)
+                {
+                    if (classSymbol == null)
+                    {
+                        continue;
+                    }
+
+                    accumulatedBytes += SerializeClass(classSymbol.Value, builder);
```

Comment:
nitpick: I recommend surrounding `SerializeClass` with try/catch that will capture ArgumentOutOfRangeException. It might be thrown if the [writing to the StringBuilder will exceed the MaxCapacity](https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=net-7.0#system-text-stringbuilder-append(system-string)).